Task2

Waltans opened the pull request.

Waltans committed:
логи

Waltans committed:
просто мр

Waltans committed:
Доработал сервисы:

- CommandService теперь возвращает Response с состоянием бота
- Добавил клавиатуру для выбора сложности пароля
- Добавил JavaDoc
- Исправил замечания
_ Поправил несколько багов

Waltans committed:
ТЕСТЫ

Waltans committed:
исправил замечания

totheskysx10 committed:
Merge pull request #3 from Waltans/task-2-suntsov

Task 2 suntsov

totheskysx10 committed:
Рефакторинг, реализация поиска и сортировки

totheskysx10 committed:
продолжение рефакторинга

totheskysx10 committed:
Рефакторинг, написание тестов

totheskysx10 committed:
Подтянул замечания из ветки 1 задания

totheskysx10 committed:
Убрал лишние логгеры и импорты

totheskysx10 committed:
Реализация хэндлеров через интерфейс, исправление замечаний

totheskysx10 committed:
Merge remote-tracking branch 'origin/task-2-surin' into task-2-surin

# Conflicts:
#	src/main/java/ru/naumen/handler/GenerateHandler.java

Waltans committed:
Добавил verify в тесты

Waltans committed:
Правил импорты

Waltans committed:
Убрал дублирование

Waltans committed:
Merge pull request #5 from Waltans/task-2.2-suntsov

totheskysx10 committed:
Merge pull request #4 from Waltans/task-2-surin

task2-surin

Waltans committed:
Fix бага

YaEtoTui commented on src/main/java/ru/naumen/bot/Command.java:
```diff
@@ -0,0 +1,69 @@
+    public static final String BY_DATE = "Дате";
+    public static final String BY_DESCRIPTION = "Описанию";
+    public static final String MENU_KEYBOARD = "Меню";
+    public static final int COMMAND_WITHOUT_PARAMS_LENGTH = 1;
```
Не используете нигде

totheskysx10 replied:
убрал

YaEtoTui commented on src/main/java/ru/naumen/bot/TelegramBot.java:
```diff
@@ -0,0 +1,169 @@
+    }
+
+    /**
+     * Клавиатура с выбором сложности
```
Ребят, методы выполняют какое-то действие. Где оно?
То начиная с этого метода и ниже. Нужно javaDoc хорошо расписывать

totheskysx10 replied:
дописал javadoc - теперь описал действие каждой кнопки

YaEtoTui commented on src/main/java/ru/naumen/exception/PasswordNotFoundException.java:
```diff
@@ -0,0 +1,10 @@
+/**
+ * Исключение, когда пароль не найден
+ */
+public class PasswordNotFoundException extends Exception {
```
Наследование от Exception? Точно от него лучше?

totheskysx10 replied:
исключение должно быть проверяемое, ведь мы ожидаем выброс этого исключения, оно выкидывается не в следствие ошибок нас (разработчиков). + так сделано по замечанию Вячеслава в 1 задаче

YaEtoTui replied:
Понял. Вопросов нет)

YaEtoTui commented on src/main/java/ru/naumen/exception/UserNotFoundException.java:
```diff
@@ -0,0 +1,10 @@
+/**
+ * Исключение, когда пользователь не найден
+ */
+public class UserNotFoundException extends Exception {
```
Наследование от Exception? Точно от него лучше?

totheskysx10 replied:
ответил выше)

YaEtoTui commented on src/main/java/ru/naumen/handler/FindHandler.java:
```diff
@@ -0,0 +1,57 @@
+import ru.naumen.service.EncodeService;
+import ru.naumen.service.PasswordService;
+
+import java.util.ArrayList;
```
импорт не используется

totheskysx10 replied:
убрал

YaEtoTui commented on src/main/java/ru/naumen/service/EncodeService.java:
```diff
@@ -0,0 +1,65 @@
+package ru.naumen.service;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
```
импорты не используются

totheskysx10 replied:
убрал

YaEtoTui commented on src/test/java/ru/naumen/bot/UserStateCacheTest.java:
```diff
@@ -0,0 +1,96 @@
+import static org.junit.jupiter.api.Assertions.*;
+
+/**
+ * ����� ��������� ������ UserStateCache
```
encoding...

YaEtoTui replied:
![Uploading изображение.png…]()

totheskysx10 replied:
исправил

YaEtoTui commented:
![изображение](https://github.com/user-attachments/assets/c80158bb-1325-4218-b81c-301d5d2ac842)

Сложность в виде чисел плохая практика, как минимум расписывают, вы то знаете, так как пишете его. А вот пользователь может не понять.
И как бы изначально кнопки были другие)

YaEtoTui commented:
![изображение](https://github.com/user-attachments/assets/960d70de-0325-4e76-8292-4ffd8d0b93fc)

Поломалась логика

YaEtoTui commented:
![изображение](https://github.com/user-attachments/assets/08a4b627-e472-4730-97e1-5e3d74e5f81d)

Не знаю в чем проблема, но пароль некорректно сохраняется :3

YaEtoTui commented:
![изображение](https://github.com/user-attachments/assets/4043f156-0942-4024-abf4-a681f928e520)

Нет, все таки это баг. Уже не первый раз, после фразы "введите описание" вот такая каша

YaEtoTui commented:
Есть замечания!

Waltans committed:
Доработка тестов

Waltans committed:
Merge pull request #7 from Waltans/task2-suntsov-fix

Fix бага

Waltans commented:
> Сложность в виде чисел плохая практика, как минимум расписывают, вы то знаете, так как пишете его. А вот пользователь может не понять.
> И как бы изначально кнопки были другие)

Фикс по багу влит

totheskysx10 committed:
bot initialize fix from 1 task

totheskysx10 committed:
bot initialize fix from 1 task

totheskysx10 committed:
Merge pull request #8 from Waltans/task-2-surin

bot initialize fix from 1 task

totheskysx10 committed:
Исправил замечания, убрал коллективную безответственность с валидацией в хэндлерах

totheskysx10 committed:
Merge remote-tracking branch 'origin/task-2-surin' into task-2-surin

totheskysx10 committed:
Исправил логическую ошибку

totheskysx10 committed:
Исправил ошибку валидации

totheskysx10 committed:
Merge pull request #9 from Waltans/task-2-surin

Task 2 surin

totheskysx10 commented:
> Сложность в виде чисел плохая практика, как минимум расписывают, вы то знаете, так как пишете его. А вот пользователь может не понять. И как бы изначально кнопки были другие)

кнопки сделал такие же, как описано в задании - теперь подписи в виде слов

totheskysx10 committed:
багофикс в тестах

totheskysx10 committed:
Merge pull request #11 from Waltans/task-2-surin

багофикс в тестах

YaEtoTui commented:
Вопросов нет. Баг поправлен.

2-ое задание принято

vpyzhyanov commented:
В пулл реквесте конфликты! Прошу их устранить

Почему первая задача не влита во вторую? У меня нет никакого желания писать одни и те же замечания, которые я писал в первой задаче. Вторая задача должна учитывать все исправления из первой.

Буду смотреть задачу только после этих исправлений

totheskysx10 committed:
Merge branch 'task1' into task2

totheskysx10 committed:
fixed merge tags in javadoc

totheskysx10 commented:
> В пулл реквесте конфликты! Прошу их устранить
> 
> Почему первая задача не влита во вторую? У меня нет никакого желания писать одни и те же замечания, которые я писал в первой задаче. Вторая задача должна учитывать все исправления из первой.
> 
> Буду смотреть задачу только после этих исправлений

Конфликты устранены, первая задача влита во вторую - теперь все изменения учитываются)

vpyzhyanov commented on src/main/java/ru/naumen/service/PasswordService.java:
```diff
@@ -152,10 +152,6 @@ public UserPassword findPasswordByUuid(String uuid) throws PasswordNotFoundExcep
 
     /**
      * Подсчитывает количество паролей пользователя
-<<<<<<< HEAD
```
ого, похоже вы мёрджите через консоль. Ну удачи вам)) Только по настоящему не ленивые программисты между простым и сложным выбирают всегда сложное)

vpyzhyanov commented on pom.xml:
```diff
@@ -13,49 +13,84 @@
-        <developerConnection/>
-        <tag/>
-        <url/>
-    </scm>
```
удивительно, что эти изменения всё ещё здесь. Может это потому, что влили не main, а task1?

Waltans replied:
Да, так и было, сейчас влили main

vpyzhyanov commented on src/main/java/ru/naumen/bot/Command.java:
```diff
@@ -0,0 +1,68 @@
+/**
+ * Команды как константы и количество допустимых параметров для каждой команды
+ */
+public class Command {
```
уместнее было использовать enum. Тогда и имя в единственном числе будет оправдано. А то сейчас описание не соответствует имени.

totheskysx10 replied:
переделал в enum

vpyzhyanov commented on src/main/java/ru/naumen/bot/Command.java:
```diff
@@ -0,0 +1,68 @@
+     * Отображение, в которой ключи - команды,
+     * значения - список допустимых количеств параметров, передаваемых вместе с командой.
+     */
+    public static final Map<String, List<Integer>> commandsAndNumberOfParams = Map.of(
```
Ребята, в чём дело? Разве так сложно соблюсти элементарные требования?

Если бы делали куда более подходящую структуру для этого класса (enum), то создавать такие поля бы не потребовалось.

totheskysx10 replied:
переделал на enum - теперь статиков нигде нет, требование соблюдается)
да, вы правы... подразумевалось, что это будут неизменяемые данные, куда статик вроде и подходит (но всё равно нарушает требования, ведь это не примитивная/стринговая константа). хотя получили совсем обратное - сейчас увидели и осознали это (и публичность ещё добавила проблем)... ошибку поняли, и поняли как сделать без статиков - всё исправлено

vpyzhyanov commented on src/main/java/ru/naumen/bot/Command.java:
```diff
@@ -0,0 +1,68 @@
+    /**
+     * Отображение, где кнопки соотносятся с командами
+     */
+    public static final Map<String, String> commandKeyMapping = Map.of(
```
ещё и всё публично. Инкапсуляция - давай до свидания. По сути весь этот класс - просто свалка статических полей, связанных с командой. Не надо так. Команда должна быть самодостаточна и должна инкапсулировать все необходимые данные внутри себя. Пока вижу только процедурное программирование

totheskysx10 replied:
переделал на enum - теперь все поля приватные, доступ через геттеры (соблюдается инкапсуляция), и все данные (команда как она есть, её подпись на клавиатуре и кол-во параметров) инкапсулируются в самих командах

vpyzhyanov commented on src/main/java/ru/naumen/bot/Constants.java:
```diff
@@ -0,0 +1,56 @@
+public class Constants {
+
+    public static final String INCORRECT_COMMAND_RESPONSE = "Введена некорректная команда! Справка: /help";
+    public static final String WELCOME_MESSAGE = "Здравствуйте. Я бот, который поможет Вам генерировать и управлять паролями.\n\n" +
```
самое время узнать что такое многострочные строки, это очень удобные штуки

totheskysx10 replied:
да, сделал вместо \n многострочные строки

vpyzhyanov commented on src/main/java/ru/naumen/bot/Constants.java:
```diff
@@ -0,0 +1,56 @@
+    public static final String PASSWORD_NOT_FOUND_MESSAGE = "Не найден пароль с id %s";
+    public static final String PASSWORD_SAVED_MESSAGE = "Пароль успешно сохранён";
+    public static final String NO_PASSWORDS_MESSAGE = "Нет ни одного пароля. Справка: /help";
+    public static final String PASSWORD_LIST_FORMAT = "\n%s) Сайт: %s, Пароль: %s";
```
в одной строчке сразу несколько проблем:

1. Если значение константы начинается или заканчивается на спецсимвол, то этот символ должен быть вынесен из этой константы. Очевидно, что это часть форматированного сообщения, но сообщения должны быть цельные и единые, а форматирование - просто их объединяет. Пусть оно объединяет их в том месте, где оно их объединяет)) Надеюсь теперь стало очевидно почему не надо спецсимволы так использовать.
2. Вижу закрывающую скобку, но не вижу открывающуюся..... хотя может так оно и задумано. Но если вдруг это число, то почему тогда вы вставляете туда строку? Вставляйте число

totheskysx10 replied:
1. про спецсимволы понятно. убрал его из константы, вынес туда, где из константы формируется непосредственно сообщение к ответу
2. скобка - это номер пароля в списке - например, "2)". сделал %d, чтобы вставлялось число

vpyzhyanov commented on src/main/java/ru/naumen/bot/Constants.java:
```diff
@@ -0,0 +1,56 @@
+    public static final String ENTER_SEARCH_REQUEST = "Введите поисковый запрос";
+    public static final String USER_NOT_FOUND = "Пользователь не найден";
+    public static final String ENCRYPT_ERROR = "Ошибка шифрования пароля";
+    public static final int EDIT_COMMAND_LENGTH_HAS_DESCRIPTION = 5;
```
1. Константы относящиеся к команде должны быть определи вместе с командой
2. Если константа используется в одном единственном месте, то она должна быть определена именно в том классе, где она используется. В данном случае в EditHandler. И быть **приватной**. Инкапсуляцию вы видимо вообще не уважаете.

Эти же замечания относятся и к другим константам

totheskysx10 replied:
константы, которые используются только один раз, вынесены туда, где они используются, сделаны приватными.
Про первый пункт, правильно понял, что вы имели в виду константу COMMAND_WITHOUT_PARAMS_LENGTH? она используется много где в проекте, поэтому выглядит логичным (как нам кажется😅) оставить её в одном месте и обращаться из разных классов, хоть она и относится непосредственно к командам. пока оставил её вместе с константами, чтобы в каждом классе не определять одно и то же. будем рады обратной связи, чтобы понять, как лучше исправить замечание :)

vpyzhyanov replied:
> Про первый пункт, правильно понял, что вы имели в виду константу COMMAND_WITHOUT_PARAMS_LENGTH?

Замечания пишутся всегда на конкретную строчку кода. В данном случае на EDIT_COMMAND_LENGTH_HAS_DESCRIPTION. Но и для некоторых других констант это замечание справедливо. Надо использовать свой ум, чтобы понять к каким и исправить там тоже.

vpyzhyanov replied:
Вот пример

![image](https://github.com/user-attachments/assets/eff9df6b-2b87-4bec-b2eb-e66c9641913e)

Остальные тоже надо исправить

totheskysx10 replied:
понял, всё исправлено) упустил момент что несколько usages могут быть в пределах одного класса😅

vpyzhyanov commented on src/main/java/ru/naumen/bot/Constants.java:
```diff
@@ -0,0 +1,56 @@
+
+    public static final String ENCRYPT_EXCEPTION = "Ошибка шифрования пароля";
+
+    public static final String DECRYPT_EXCEPTION = "Ошибка дешифрования пароля";
```
2 этих константы нигде не используются

totheskysx10 replied:
теперь их нет)

vpyzhyanov commented on src/main/java/ru/naumen/bot/Constants.java:
```diff
@@ -0,0 +1,56 @@
+/**
+ * Константы
+ */
+public class Constants {
```
не надо устраивать свалку из констант. Как минимум они должны быть поделены по темам. И не надо в общую кучу выкидывать константы, которые нужны только в одном месте.

totheskysx10 replied:
константы разделил по темам, используемые один раз - вынес и сделал приватными

vpyzhyanov replied:
>  используемые один раз - вынес и сделал приватными

С этим почему-то возникли проблемы. Ещё один пример:

![image](https://github.com/user-attachments/assets/02412d1f-597a-4fae-be91-12289b15360d)

totheskysx10 replied:
исправлено

vpyzhyanov commented on src/main/java/ru/naumen/bot/Response.java:
```diff
@@ -0,0 +1,6 @@
+
+import ru.naumen.model.State;
+
+public record Response(String message, State botState) {
```
Ребята, в чём дело? Почему возникают такие сложности, почему не получается соблюсти элементарные требования?

Пока не приведёте код в соответствие требованиям, дальше проверять не буду!

totheskysx10 replied:
да, упущен момент с джавадоком... дописал его. тщательно проверили весь проект на наличие джавадоков и отсутствие статиков кроме констант -> код приведён в соответствие требованиям

vpyzhyanov reviewed (changes requested):
Ребята, в чём дело? Почему возникают такие сложности, почему не получается соблюсти элементарные требования?

Пока не приведёте код в соответствие требованиям, дальше проверять не буду!

vpyzhyanov commented:
> Конфликты устранены, первая задача влита во вторую - теперь все изменения учитываются)

Почему-то "левые" изменения всё ещё остались. Например изменения pom файла.

что интересно, я попробовал влить main в task2 и получился 45 изменённых файлов.😳 Как так? Я ожидал, что не будет ни одного

totheskysx10 committed:
Исправление замечаний

totheskysx10 committed:
Merge branch 'task2' into task-2-surin

# Conflicts:
#	src/main/java/ru/naumen/bot/Constants.java

totheskysx10 committed:
Исправление замечаний - javadoc

totheskysx10 committed:
Исправление замечаний - однострочное программирование

totheskysx10 committed:
Исправление замечаний - javadoc

Waltans committed:
Merge branch 'refs/heads/main' into task2

totheskysx10 committed:
доработал константы

totheskysx10 committed:
Merge pull request #16 from Waltans/task-2-surin

Исправление замечаний

Waltans commented:
> Почему-то "левые" изменения всё ещё остались. Например изменения pom файла.
> 
> что интересно, я попробовал влить main в task2 и получился 45 изменённых файлов.😳 Как так? Я ожидал, что не будет ни одного

теперь при вливании main в task2 нет изменений

totheskysx10 committed:
убрал статик в тестах

totheskysx10 committed:
Merge pull request #18 from Waltans/task-2-surin

убрал статик в тестах

totheskysx10 committed:
Избавился от маппера - тёмной стороны

totheskysx10 committed:
Merge pull request #19 from Waltans/task-2-surin

Избавился от маппера - тёмной стороны

totheskysx10 committed:
фикс названия - соответствие ООП

totheskysx10 committed:
javadoc

totheskysx10 committed:
Merge pull request #20 from Waltans/task-2-surin

фикс названия - соответствие ООП

totheskysx10 committed:
убрал статик импорт стейта

totheskysx10 committed:
Merge pull request #21 from Waltans/task-2-surin

убрал статик импорт стейта

totheskysx10 committed:
убрал лишнюю проверку на null

totheskysx10 committed:
Merge pull request #22 from Waltans/task-2-surin

убрал лишнюю проверку на null

vpyzhyanov commented on src/main/java/ru/naumen/bot/TelegramBot.java:
```diff
@@ -37,7 +47,7 @@ public TelegramBot(@Value("${bot.token}") String botToken,
     @EventListener({ContextRefreshedEvent.class})
     public void initialize() {
-        try {
+        try{
```
зачем? Почему не используете форматирование? Я уже рассказывал как сделать это автоматически при коммите

Waltans replied:
Не за чем. Исправил

vpyzhyanov commented on src/main/java/ru/naumen/bot/Response.java:
```diff
@@ -0,0 +1,11 @@
+/**
+ * Ответ бота
+ * @param message сообщение с ответом
+ * @param botState состояние бота после ответа
```
Ну вот и открылась тёмная сторона. Уже становится очевидно, что две логически не связанные друг с другом функциональности объединены в один класс. Непонятно зачем.

Вопрос для самоконтроля: каким образом "состояние бота" является часть ответа? Вы серьёзно его пользователю передаёте? Пользователь меняет состояние?

Waltans replied:
Наверное, это скорее не самый удачный javaDoc, тут скорее состояние пользователя после ответа бота, бот никакого отношения не имеет к состоянию ответа. Да, мы передаем состояние, только не пользователю, а боту для обработки. Пользователь меняет свое состояние.

Но справедливости ради, замечу, что состояние пользователя не обязательно передавать с ответом, мы всегда его можем узнать получив из кеша, поэтому в какой-то мере это избыточно, уберу это и переделаю обработку команд ботом

vpyzhyanov replied:
> его можем узнать получив из кеша, поэтому в какой-то мере это избыточно

Это не просто избыточно, это категорически неправильно. Но теперь стало правильно, теперь сама команда меняет состояние, обращаясь к соответствующему сервису. Так и должно было быть с самого начала. То есть смена состояния - это ответственность именно обработчика конкретной команды.

vpyzhyanov commented on src/main/java/ru/naumen/bot/TelegramBot.java:
```diff
@@ -58,30 +68,137 @@ public void onUpdateReceived(Update update) {
+        replyKeyboardMarkup.setOneTimeKeyboard(true);
+
+        if (response.botState().equals(State.NONE)) {
+            List<KeyboardRow> keyboardRows = mainKeyboard();
```
И ведь действительно. Какое-то состояние передали и накрутили логику, который тут быть вообще не должно. **Добро пожаловать в коллективную безответственность**. В одночасье обесценили класс CommandService отжав у него часть работы.

Как правильно: Клавиатура должна определять в классе, отвечающем за логику. Про "состояния" вообще никто, коме самой логики, знать не должен. Класс Response должен передавать информацию по которой телеграмм бот сможет построит произвольную клавиатуру.

Waltans replied:
Теперь нет такой логики, комманд сервис решает какую клавиатуру показать пользователю

vpyzhyanov replied:
Ай яй яй, сделали страшное, затянули зависимость от телеграмма в commandService. Притом сделав это очень не очевидно. Прочувствуйте: метод отправки сообщения в чат обращается в какой-то сервис по **обработке** команд, чтобы получить клавиатуру. Тут возникает масса вопросов: 

1. как он узнает какую клавиатуру надо сейчас отобразить? 
2. То есть клавиатура вообще не связана с обработкой команды? Но это же не правда!

Самый правильный способ - добавить это в класс ответа. Вероятно потребуется создать свой класс клавиатуры.

Waltans replied:
Добавл класс клавиатуры и в ответ добавил клавиатуру

vpyzhyanov commented on src/main/java/ru/naumen/bot/TelegramBot.java:
```diff
@@ -58,30 +68,137 @@ public void onUpdateReceived(Update update) {
+     * Средний (COMPLEXITY_MEDIUM),
+     * Сложный (COMPLEXITY_HARD)
+     */
+    private List<KeyboardRow> complexityKeyBoard() {
```
тут не место этому методу

Waltans replied:
Вынес методы

vpyzhyanov replied:
Стало чуть лучше, но исходный посыл замечания не исправлен. Связь с телеграммом как была, так и осталась.

Waltans replied:
Убрал связь с телеграмом

vpyzhyanov commented on src/main/java/ru/naumen/bot/TelegramBot.java:
```diff
@@ -58,30 +68,137 @@ public void onUpdateReceived(Update update) {
+     * Клавиатура с выбором типа сортировки
+     * Можно выбрать по дате (BY_DATE) и описанию (BY_DESCRIPTION)
+     */
+    private List<KeyboardRow> sortKeyBoard() {
```
О_о вот это поворот. Если вы пытались ввести в заблуждение, то у вас это отлично получилось. Читаем название метода "Сортировать клавиатуру" (хотя на самом деле надо прочитать как "Сортировать ключ доски", т.е. keyboard - это одно слово).

Правильное название метода `createSelectSortTypeKeyboard`

Waltans replied:
Исправил названия методов

vpyzhyanov commented on src/main/java/ru/naumen/bot/TelegramBot.java:
```diff
@@ -58,30 +68,137 @@ public void onUpdateReceived(Update update) {
+     * SORT - отсортировать пароли
+     * FIND - поиск паролей по описанию
+     */
+    private List<KeyboardRow> listKeyBoard() {
```
похоже все методы так или иначе вводят в заблуждение, т.к. названы неправильно.

Запомните простое правило: метод **ВСЕГДА** **глагол!**

Waltans replied:
Да, понятно

vpyzhyanov commented on src/main/java/ru/naumen/bot/UserStateCache.java:
```diff
@@ -0,0 +1,91 @@
+package ru.naumen.bot;
```
опять заглянули на тёмную сторону. Какое отношение кеш имеет к боту? Не увидел ни одной связи с ботом

Waltans replied:
перенес класс. Отношение к боту не имеет

vpyzhyanov commented on src/main/java/ru/naumen/bot/UserStateCache.java:
```diff
@@ -0,0 +1,91 @@
+            return totalUserParams.get(userId);
+        } else {
+            totalUserParams.put(userId, new ArrayList<>());
+            return new ArrayList<>();
```
привыкайте работать с неизменяемыми коллекциями. `List.of()`

Waltans replied:
сделал

vpyzhyanov commented on src/main/java/ru/naumen/service/ValidationService.java:
```diff
@@ -0,0 +1,169 @@
+        String commandString = splitCommand[0];
+        int paramsCount = splitCommand.length - 1;
+
+        State state = userStateCache.getUserState(userId);
```
подозрение на коллективную безответственность. Сильно подозреваю, что класс userStateCache лишён инкапсуляции и очень зря ходит к кому попало.

Что мы имеем, "Сервис валидации введённых данных" для проверки корректности команды требует указать ID пользователя. То есть вы хотите сказать проверка команды зависит от того, кто её ввёл?! Это не может быть так. Я ещё могу понять, если ValidationService зависит от состояния (и то не очень понимаю), но от пользователя точно не должен зависеть.

Сильное подозрение, что ValidationService ворует часть работы у обработчиков команды. Всё-таки с состоянием обычно работают именно они и больше никто не должен работать с состоянием

Waltans replied:
Теперь этого класса нет

vpyzhyanov commented on src/main/java/ru/naumen/service/ValidationService.java:
```diff
@@ -0,0 +1,169 @@
+     * @return корректна ли сложность
+     */
+    public boolean isValidComplexity(String complexity) {
+        return complexity.equals("1") ||
```
для лёгкости чтения кода, логическое условие нужно писать в начале строки

Waltans replied:
ок

vpyzhyanov commented on src/main/java/ru/naumen/service/ValidationService.java:
```diff
@@ -0,0 +1,169 @@
+     * @param splitCommand - разделенный список параметров
+     * @return - true, если все параметры удовлетворяют
+     */
+    public boolean areNumbersEditCommandParams(String[] splitCommand) {
```
инкапсуляция вышла и не вернулась. Метод обязан быть private. А то, что тесты его используют меня совершенно не волнует.

Waltans replied:
Убрал методы

vpyzhyanov commented on src/main/java/ru/naumen/service/ValidationService.java:
```diff
@@ -0,0 +1,169 @@
+     * @param splitCommand - разделенный список параметров
+     * @return - true, если все параметры удовлетворяют
+     */
+    public boolean areNumbersDeleteCommandParams(String[] splitCommand) {
```
все методы ниже аналогично. Почему-то не дружите с инкапсуляцией

Waltans replied:
убрал

vpyzhyanov commented on src/main/java/ru/naumen/bot/command/CommandFinder.java:
```diff
@@ -0,0 +1,28 @@
+     * @return команда
+     * @throws CommandNotFoundException когда команда не найдена
+     */
+    public Command findCommand(String input) throws CommandNotFoundException {
```
1. Слишком жирно создавать целый класс для такой простой задачи. CommandService сам в состоянии разобраться со своими командами, всё-таки это его задача вызвать нужную команду. Привет коллективной безотстветственности.
2. Линейная сложность? Не, так не пойдёт. Вы разве не изучали коллекции? Самое время использовать полученные знания

В общем класс надо удалить за ненадобностью. А если вы переживаете за ValidationService, то не нужно переживать, он вообще сомнительный персонаж. Вообще-то каждая команда сама себя провалидировать. И правила валидации у каждой из них могут быть **сильно разные**, нельзя их объединять в одну гребёнку.

Waltans replied:
Жирно, занес логику в commandService. Переделал на hashmap и получим константную сложность. Удалил класс

vpyzhyanov replied:
Ну вот, другое дело!

vpyzhyanov commented on src/main/java/ru/naumen/service/ValidationService.java:
```diff
@@ -0,0 +1,169 @@
+ * Сервис валидации введённых данных
+ */
+@Service
+public class ValidationService {
```
на всякий случай продублирую

> А если вы переживаете за ValidationService, то не нужно переживать, он вообще сомнительный персонаж. Вообще-то каждая команда сама себя провалидировать. И правила валидации у каждой из них могут быть **сильно разные**, нельзя их объединять в одну гребёнку.

Waltans replied:
убрал validationSerivice

vpyzhyanov commented on src/main/java/ru/naumen/service/CommandService.java:
```diff
@@ -72,248 +65,39 @@ public String performCommand(String message, long userId, String username) {
+            CommandHandler handler = commandHandlers.get(command.getCommand());
+            return handler.handle(splitCommand, userId);
+        } catch (CommandNotFoundException e) {
+            return performNotCommandMessage(splitCommand, userId);
```
в этот раз у вас не получилось меня обмануть! Сначала вы ищете команду. Почему-то утверждаете, что её отсутствие - это некорректная ситуация (исключение бросается только в некорректной ситуации).

![image](https://github.com/user-attachments/assets/ec9e92dc-61ef-4bb1-8937-29aa24b00dcd)

А дальше как ни в чём не бывало продолжаете обработку.

В общем убираем исключение. Ситуация корректна. Метод findCommand должен возвращать Optional

Waltans replied:
Исправил ситуацию

vpyzhyanov commented on src/main/java/ru/naumen/bot/command/CommandFinder.java:
```diff
@@ -0,0 +1,28 @@
+                return cmd;
+            }
+        }
+        throw new CommandNotFoundException("Команда не найдена: " + input);
```
дублирую

> убираем исключение. Ситуация корректна. Метод findCommand должен возвращать Optional

Waltans replied:
Больше его нет

vpyzhyanov commented on src/main/java/ru/naumen/bot/command/Command.java:
```diff
@@ -0,0 +1,115 @@
+    Command(String command, String keyboardLabel, List<Integer> validParamCounts) {
+        this.command = command;
+        this.keyboardLabel = keyboardLabel;
+        this.validParamCounts = validParamCounts;
```
теперь я понял что не так. Вы попытались унифицировать неунифицируемые вещи. Такая попытка обречена на провал, а пострадает ООП.

Waltans replied:
Хм, наверное, так и есть

totheskysx10 replied:
сделали валидацию в самих обработчиках

vpyzhyanov commented on src/test/java/ru/naumen/service/ValidationServiceTest.java:
```diff
@@ -0,0 +1,203 @@
+    @Test
+    void testAreNumbersDeleteCommandParams_InvalidNumber() {
+        String[] splitCommand = {"/del", "one"};
+        Assertions.assertFalse(validationService.areNumbersDeleteCommandParams(splitCommand));
```
тестирование методов, которые вообще в проекте не используются

totheskysx10 replied:
удалены, как и сам validationService

vpyzhyanov reviewed (changes requested):
слишком много тёмной стороны

Waltans committed:
Исправление ошибок

- Разнес ValidationService.java, нужно ещё править тесты
- Удалил ненужные классы
- Почистил методы
- Доработал тесты

Waltans committed:
Поменял валидацию, внес в хендлеры

Waltans committed:
Еще немного логики поменял в валидаторах и с клавиатурой прикололся

Waltans committed:
Еще немного логики поменял в валидаторах и с клавиатурой прикололся

Waltans committed:
Merge pull request #23 from Waltans/task2-suntsov

Исправление ошибок

totheskysx10 committed:
Исправление замечаний

totheskysx10 committed:
Исправление валидации индекса

totheskysx10 committed:
Merge pull request #25 from Waltans/task-2-surin

Исправление замечаний task2

vpyzhyanov commented on src/main/java/ru/naumen/service/CommandService.java:
```diff
@@ -20,84 +25,104 @@
-        } catch (CommandNotFoundException e) {
-            return performNotCommandMessage(splitCommand, userId);
-        }
+    public List<KeyboardRow> getKeyboards(long userId) {
```
> Ай яй яй, сделали страшное, затянули зависимость от телеграмма в commandService. Притом сделав это очень не очевидно. Прочувствуйте: метод отправки сообщения в чат обращается в какой-то сервис по обработке команд, чтобы получить клавиатуру

Waltans replied:
да, понятна ошибка

vpyzhyanov commented on src/main/java/ru/naumen/service/CommandService.java:
```diff
@@ -1,15 +1,20 @@
-import ru.naumen.bot.command.Command;
-import ru.naumen.bot.command.CommandFinder;
+import org.springframework.transaction.annotation.Transactional;
+import org.telegram.telegrambots.meta.api.objects.replykeyboard.buttons.KeyboardRow;
```
вот тут очевидна лишняя зависимость

Waltans replied:
fix

vpyzhyanov reviewed (approved):
Как исправите последние замечания, можете мёрджить. Задача принята.

Оценка:

- ООП: 4 балла (снизил за то, что переметнулись на тёмную сторону, т.е. прокинули связь с телеграммом внутрь обработчика команд)
- GIT: 3 балла

Waltans committed:
Response передает клавиатуру, какую надо установить сервису

Waltans committed:
fix

Waltans committed:
перенес класс клавиатуры

Waltans committed:
javaDoc

Waltans committed:
убрал метод

Waltans committed:
fix

Waltans committed:
fix

Waltans committed:
Merge pull request #32 from Waltans/task2-suntsov-keyboard-service

Response передает клавиатуру, какую надо установить сервису
